components/raftstore-v2/src/operation/command/mod.rs

```diff
@@ -147,19 +206,160 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
                     // on_committed callback if necessary.
                     p.cb.notify_committed();
                 }
-                entry_and_proposals.push((e, proposal));
+                entry_and_proposals.push((e, proposal.map_or_else(Vec::new, |p| p.cb)));
             }
         } else {
-            entry_and_proposals = committed_entries.into_iter().map(|e| (e, None)).collect();
-        }
-        // Note that the `commit_index` and `commit_term` here may be used to
-        // forward the commit index after being restarted. So it must be less
-        // than or equal to persisted index.
-        let commit_index = cmp::min(
-            self.raft_group().raft.raft_log.committed,
-            self.raft_group().raft.raft_log.persisted,
-        );
-        let commit_term = self.raft_group().raft.raft_log.term(commit_index).unwrap();
-        // TODO: schedule apply task
+            entry_and_proposals = committed_entries.into_iter().map(|e| (e, vec![])).collect();
+        }
+        // Unlike v1, v2 doesn't need to persist commit index and commit term. The
+        // point of persist commit index/term of raft apply state is to recover commit
+        // index when the writes to raft engine is lost but writes to kv engine is
+        // persisted. But in v2, writes to raft engine must be persisted before
+        // memtables in kv engine is flushed.
+        let apply = CommittedEntries {
+            entry_and_proposals,
+        };
+        self.apply_scheduler()
+            .send(ApplyTask::CommittedEntries(apply));
+    }
+
+    pub fn on_apply_res(&mut self, apply_res: ApplyRes) {
+        if !self.serving() {
+            return;
+        }
+        // It must just applied a snapshot.
+        if apply_res.applied_index < self.entry_storage().first_index() {
+            // TODO: handle admin side effects like split/merge.
+            return;
+        }
+        self.raft_group_mut()
+            .advance_apply_to(apply_res.applied_index);
+        let is_leader = self.is_leader();
+        let entry_storage = self.entry_storage_mut();
+        entry_storage
+            .apply_state_mut()
+            .set_applied_index(apply_res.applied_index);
+        entry_storage.set_applied_term(apply_res.applied_term);
+        if !is_leader {
+            entry_storage.compact_entry_cache(apply_res.applied_index + 1);
+            // TODO: handle read.
```

Comment:
Question: why need to handle read here?  Do you mean the empty entry in read index when transferring leader?

Reply:
There may be reads are paused due to split/merge/new leaders.

Reply:
> "After read index, it may wait for application to respond."
What's the exact respond need to be done here? Is v1's behavior same?

Reply:
The same. Just check if log has applied to the corresponding index, if it does, respond to read.